codec-http/src/main/java/io/netty/handler/codec/http/multipart/DefaultHttpDataFactory.java

```diff
@@ -321,7 +321,9 @@ public void cleanAllHttpData() {
 
             List<HttpData> list = e.getValue();
             for (HttpData data : list) {
-                data.release();
+                if (data.refCnt() > 0) {
```

Comment:
I think this is not really correct. We should release if always. If there is an exception it is a bug.

Reply:
I agree and I will revert this.

I redo my analysis and I understand my mistake while trying to fix the issue raised during the test on Github.

If I call `data.release()` (on FileUpload item) before `cleanAllHttpData()`:
- In Memory: OK final refCnt is 0 on `undecodedChunk` 
- On Disk : OK final refCnt is 0 on `undecodedChunk` 

If I only call `cleanAllHttpData()`: 
- In Memory: the FileUpload item is not released (`undecodedChunk` keeps all refCnt)
   -  This is where the issue was raised during tests.
- On Disk: OK final refCnt is 0 on `undecodedChunk` 

According to method definition of `cleanAllHttpData()`, this is normal since it only concerns disk based attributes.
So you are perfectly right, this is not a good solution.
So I remove this and I fix the test according to correct usage (calling release method on each HttpData).

Reply:
I think we should try to release if possible during destroy (Disk based is done this way too, but not Memory based) to be consistent.
But in that case, if one mixed release of HttpData, cleanAllHttpData and destroy will raised an exception on refCnt during cleanAllHttpData (since object is already released unitary).

So I propose a new solution, still keeping this test (already there in destroy to prevent such an issue but on undecodedChunk only).

@normanmaurer WDYT?

Reply:
This solution is then homogeneous between Disk, Mixed and Memory based factories of course (same behaviour).

Reply:
Note that either letting at it was or changing it brings an issue:
- keeping as it was: Disk based and Memory based have not the same behaviiour when releasing items, and the issue is then what to do when in Mixed mode
- changing it: they will behave the same but Memory based will lead to "virtual" lost of memory based items after destroying the decoder, when it was not the case before (the user shall call it when needed)

I'm a bit in the middle:
- making the two acting the same way
- chaning a behaviour which could break some existing codes

My proposal would be: (not the current code)
- not changing behaviour (so not having the Memory based cleaning HttpData by calling forcingly the reslease method (removing this)
- but keeping refCnt guards such as, at least, no exception are raised while release is beeing called under the wood already (disk based mainly), so minor changes, just as the one you peek (keeping the other changes)

@normanmaurer WDYT?

Reply:
@normanmaurer I choose, for now until other notice, that my new proposal could be a good tradeof:
- keeping current behaviours, even if not really the same on both sides, in order to keep the current existing client codes functional
- adding the relevant guards when releasing buffers cince they are not equivalent everywhere but will tend to allow the same kind of usage, whatever Memory, Mixed or Disk based, and believing that thos guards prevent errors that are not "client" faults, but underlying implementations that are not aligned to each other (mainly Disk and Memory based)

Reply:
@normanmaurer I guess that I leave as current code is?

Behaviours are differents between Disk and Memory based Factories, but this small fix allows to at least prevent exception to be thrown when a user release twice a buffer with the "almost" impossibility to know why (in particular in Mixed mode, where the user will have to choose the way to release it according to its real status, Disk or Memory based).

Fixing the different behavious would lead to a major change in the contract for long in place (Disk based releases virtually all underlying buffers/files, while Memory based does not except if calling explicitely `HttpData.release()`). Without this "guard", one can have refCnt exception in Mixed mode while trying to release each HttpData and then destroying the decoder (which brings the destruction of HttpData based on disk, giving the exception).

I believe we should not change the behaviour, but we can prevent internal exception due to two different implementations with this simple guard.